src/transform/src/analysis.rs

```diff
@@ -758,3 +794,1000 @@ mod non_negative {
+
+                // ALL unique columns in this equivalence
+                if all_unique {
+                    // these inputs have unique keys for _all_ of the equivalence, so they're a bound on how many rows we'll get from those sources
+                    // we'll find the leftmost such input and use it to hold the minimum; the other sources we set to 1.0 (so they have no effect)
+                    let mut sources = unique_sources.iter();
+
+                    let lhs_idx = *sources.next().unwrap();
+                    let mut lhs = std::mem::replace(&mut inputs[lhs_idx], SymExp::f64(1.0));
+                    for &rhs_idx in sources {
+                        let rhs = std::mem::replace(&mut inputs[rhs_idx], SymExp::f64(1.0));
+                        lhs = SymExp::min(lhs, rhs);
+                    }
+
+                    inputs[lhs_idx] = lhs;
+
+                    // best option! go look at the next equivalence
+                    continue;
+                }
+
+                // some unique columns in this equivalence
+                for idx in unique_sources {
+                    // when joining R and S on R.x = S.x, if R.x is unique and S.x is not, we're bounded above by the cardinality of S
+                    inputs[idx] = SymExp::f64(1.0);
+                }
+            }
+
+            SymbolicExpression::product(inputs)
+        }
+
+        fn reduce(
+            &self,
+            group_key: &Vec<MirScalarExpr>,
+            expected_group_size: &Option<u64>,
+            input: &SymExp,
+        ) -> SymExp {
+            // TODO(mgree): if no `group_key` is present, we can do way better
+
+            if let Some(group_size) = expected_group_size {
+                input / f64::cast_lossy(*group_size)
+            } else if group_key.is_empty() {
+                SymExp::from(1.0)
+            } else {
+                // in the worst case, every row is its own group
+                input.clone()
+            }
+        }
+
+        fn topk(
+            &self,
+            group_key: &Vec<usize>,
+            limit: &Option<MirScalarExpr>,
+            expected_group_size: &Option<u64>,
+            input: &SymExp,
+        ) -> SymExp {
+            // TODO: support simple arithmetic expressions
+            let k = limit
+                .as_ref()
+                .and_then(|l| l.as_literal_int64())
+                .map_or(1, |l| std::cmp::max(0, l));
+
+            if let Some(group_size) = expected_group_size {
+                input * (f64::cast_lossy(k) / f64::cast_lossy(*group_size))
+            } else if group_key.is_empty() {
+                SymExp::from(k)
+            } else {
+                // in the worst case, every row is its own group
+                input.clone()
+            }
+        }
+
+        fn threshold(&self, input: &SymExp) -> SymExp {
+            // worst case scaling factor is 1
+            input.clone()
+        }
+    }
+
+    impl Analysis for Cardinality {
```

Comment:
Not covered?

Reply:
Yes, but  this is also not yet used in production.